cmd/kubeadm/app/cmd/phases/init/uploadcerts.go

```diff
@@ -0,0 +1,114 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"github.com/pkg/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	bootstraputil "k8s.io/cluster-bootstrap/token/util"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
+	cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	nodebootstraptokenphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/bootstraptoken/node"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/uploadcerts"
+)
+
+type uploadCertsData interface {
+	Client() (clientset.Interface, error)
+	UploadCerts() bool
+	CertificateKey() string
+	CertificateDir() string
+}
+
+// NewUploadCertsPhase returns the uploadCerts phase
+func NewUploadCertsPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "upload-certs",
+		Short: "Upload certificates to kubeadm-certs",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Run:   runUploadCerts,
+		InheritFlags: []string{
+			options.CfgPath,
+			options.UploadCerts,
+			options.CertificateKey,
+		},
+	}
+}
+
+func runUploadCerts(c workflow.RunData) error {
+	data, ok := c.(uploadCertsData)
+	if !ok {
+		return errors.New("upload-certs phase invoked with an invalid data struct")
+	}
+
+	//skip upload certs.
+	if !data.UploadCerts() {
+		return nil
+	}
+	client, err := data.Client()
+	if err != nil {
+		return err
+	}
+
+	// Create the short lived bootstrap token
+	tokens, err := generateShortLivedToken()
+	if err != nil {
+		return errors.Wrap(err, "error generating token to upload certs")
+	}
+	if err := nodebootstraptokenphase.CreateNewTokens(client, tokens); err != nil {
+		return errors.Wrap(err, "error creating token")
+	}
+
+	if err := uploadcerts.UploadCerts(client, data.CertificateDir(), certsToUpload(), tokens[0]); err != nil {
+		return errors.Wrap(err, "error uploading certs")
+	}
+	return nil
+}
+
+func generateShortLivedToken() ([]kubeadmapi.BootstrapToken, error) {
+	tokenStr, err := bootstraputil.GenerateBootstrapToken()
+	if err != nil {
+		return nil, err
+	}
+	token, err := kubeadmapi.NewBootstrapTokenString(tokenStr)
+	if err != nil {
+		return nil, err
+	}
+	return []kubeadmapi.BootstrapToken{{
+		Token: token,
+		TTL: &metav1.Duration{
+			Duration: kubeadmconstants.DefaultCertTokenDuration,
+		},
+	}}, nil
+}
+
+func certsToUpload() []string {
+	return []string{
+		kubeadmconstants.CACertName,
+		kubeadmconstants.CAKeyName,
+		kubeadmconstants.FrontProxyCACertName,
+		kubeadmconstants.FrontProxyCAKeyName,
+		kubeadmconstants.ServiceAccountPublicKeyName,
+		kubeadmconstants.ServiceAccountPrivateKeyName,
+		kubeadmconstants.EtcdCACertName,
```

Comment:
If using external etcd, it is necessary to copy all the etcd certs defined in the ExternalEtcd struct (`CAFile`, `CertFile`, `KeyFile`); if we are using LocalEtcd instead it is ok to copy `EtcdCACert` and `EtcdCAKey` as proposed